api/src/pcapi/alembic/versions/20240408T150954_0fdfda2e6800_add_price_to_allocine_venue_provider.py

```diff
@@ -0,0 +1,30 @@
+"""Add price to AllocineVenueProvider (1/2)
+"""
+
+from alembic import op
+import sqlalchemy as sa
+
+
+# pre/post deployment: pre
+# revision identifiers, used by Alembic.
+revision = "0fdfda2e6800"
+down_revision = "cef3565151cc"
+branch_labels: tuple[str] | None = None
+depends_on: list[str] | None = None
+
+
+def upgrade() -> None:
+    op.add_column("allocine_venue_provider", sa.Column("price", sa.Numeric(precision=10, scale=2), nullable=True))
+    op.execute(
+        """
+        UPDATE allocine_venue_provider
+        SET price = allocine_venue_provider_price_rule.price
+        FROM allocine_venue_provider_price_rule
+        WHERE allocine_venue_provider.id = allocine_venue_provider_price_rule."allocineVenueProviderId";
+        """
+    )
+    op.alter_column("allocine_venue_provider", "price", existing_type=sa.NUMERIC(precision=10, scale=2), nullable=False)
```

Comment:
On ne peut pas établir cette contrainte `NOT NULL` dans une migration **pré**-déploiement. On aurait alors des erreurs dans le cas suivant : 
1. Les migrations pré-déploiement sont jouées. La colonne est `NOT NULL`.
2. Avant le déploiement de la nouvelle version du code, on tente d'ajouter un `AllocineVenueProvider`. Le code n'ayant pas connaissance de la colonne `price`, SQLAlchemy demande d'ajouter une ligne sans valeur pour cette colonne. PostgreSQL refusera.

L'ajout de cette contrainte doit donc s'effectuer dans une migration **post-**déploiement.

D'autre part (et même si on déplace l'ajout de cette contrainte), il y a quand même un trou quelque part avec le scénario suivant : 

1. Les migrations pré-déploiement sont jouées. La colonne existe et est nullable. Elle est remplie pour les lignes existantes.
2. Avant le déploiement de la nouvelle version du code, on joute un `AllocineVenueProvider`. Le code n'ayant pas connaissance de la colonne `price`, la valeur est `NULL`.
3. Le nouveau code est déployé et utilise désormais `AllocineVenueProvider.price`. La valeur est `NULL`, ce qui n'est probablement pas très bien pris en charge par le code. ;)

Je n'ai pas trop creusé, mais je ne pense pas qu'il y ait une solution à ce problème sans avoir une migration en 2 étapes (et donc 2 déploiements) : 

1. Une 1ère étape ajouterait la colonne (en **pré**-déploiement), remplirait la colonne dans une migration **post**-déploiement, et ferait de la double écriture : à la fois dans la nouvelle `AllocineVenueProvider.price` et dans l'actuelle `AllocineVenueProviderPriceRule`.
2. Une 2ème étape (lors d'une prochaine MEP) modifierait le code pour arrêter d'écrire et de lire dans AllocineVenueProvider.price` et lirait/écrirait exclusivement dans `AllocineVenueProvider.price`, et supprimerait la colonne (dans une migration **post-**déploiement).

Reply:
La stratégie multi-déploiement est claire pour moi sauf le passage de la colonne `AllocineVenueProvider.price` de nullable à non-nullable. Ce passage doit-il être fait lors du 2ème déploiement dans la migration pre ?

Reply:
La colonne ne peut pas être nullable avant que la double-écriture ait commencé (donc après le déploiement du code de la 1ère étape). Donc dans dans la migration **post-**déploiement de la 1ère étape, après le remplissage initial de `AllocineVenueProvider.price` à partir des `AllocineVenueProviderPriceRule` existants (qui doit aussi se faire dans une migration **post-**déploiement).

Reply:
A cause `Check database and model are aligned`, il n'est pas possible d'avoir un commit avec : 
- le code laissant `AllocineVenueProvider.price` nullable, en attendant le prochain déploiement
- une migration post qui rend `AllocineVenueProvider.price` non-nullable

Dans ce cas, je propose de rendre la colonne en base de donnée non-nullable dans la migration `pre` du prochain déploiement

Reply:
Comme discuté entre nous : bien vu, erreur de ma part. J'approuve ta solution.